core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -126,15 +138,12 @@ protected ManifestReader(
       specId = Integer.parseInt(specProperty);
     }
 
-    if (specsById != null) {
-      this.spec = specsById.get(specId);
+    if (specsById != null && specsById.containsKey(specId)) {
```

Comment:
Not sure why we need to add this  extra case here?  This spec id is for sure read from file (not the one you passed in), so all assumptions should still hold right?

Reply:
Yes, this part of the code is the same as the above part. Make sure we could get the spec from the `specId` and `specsById`. Here we check again because the `specId` is read from the file.

Reply:
I see it now.  Yea I'm not entirely sure we need all these checks if specsById contains specId, now unfortunately it gets really complicated.  But I dont have the background to say we don't need it.

I do notice here, there is a bit inefficiency in case specsById == null.  Here you will read specId from file.  It's  unnecessarily , as there is no map to look it up, instead it should go directly to read the spec from file.  In previous code it would go straight to read spec from file.

Your code:
```
if (specsById != null && specsById.contains(specId)) {
   return specsById.get(specId)
else
   specId = readSpecId()
   if (specsById != null && specsById.contains(specId)
      return specById.get(specId)
```

More optimal code:
```
if (specsById == null)
   return readSpec

if (specsById.contains(specId)
   return specsById.get(specId)
else 
   specId = readSpecId()
   if (specsById.contains(specId)
      return specsById.get(specId)
   else
      return readSpec()
```


I hope I understand it right.

Reply:
Updated the code, I think the most concerns come from the guard code: `specsById.contains(specId)`. I removed it and match the existing implementation.